### PR TITLE
Add Bitbucket Data Center support

### PR DESCRIPTION
**Pull request by zainq11**
## Description

On Feb 15, 2024, Atlassian stopped [support](https://bitbucket.org/blog/cloud-migration-benefits) for Bitbucket Server. This PR introduces changes in go-tfe so that we can support Bitbucket Data Center while keeping Bitbucket Server. The docs, tests and examples have been updated to reference Bitbucket Data Center instead of Bitbucket Server to align with the deprecation of server.

_Remember to:_

- [X] _Update the [Change Log](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md)_
- [X] _Update the [Documentation](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md#updating-the-documentation)_

## Testing plan

1.  Running CI
2.  local testing with updated terraform provider creates a bitbucket data center vcs connection.

```hcl
resource "tfe_oauth_client" "test" {
  name = "bbdc-cli-tester"
  # organization     = "zain-tfe"
  api_url          = "https://bitbucket.dev.tf-vcs.hashicorp.services"
  http_url         = "https://bitbucket.dev.tf-vcs.hashicorp.services"
  key              = "my_key"
  secret           = "-----BEGIN RSA PRIVATE KEY-----XXXXX-----END RSA PRIVATE KEY-----\n"
  rsa_public_key   = "-----BEGIN PUBLIC KEY-----XXXXX-----END PUBLIC KEY-----\n"
  service_provider = "bitbucket_data_center"
}
```

image

## External links

_Include any links here that might be helpful for people reviewing your PR. If there are none, feel free to delete this section._

- [go-tfe documentation](https://pkg.go.dev/github.com/hashicorp/go-tfe?tab=doc#xxxx)
- [go-tfe PR](https://github.com/hashicorp/go-tfe/pull/879)

## Output from acceptance tests

Successfully ran all tests in `resource_tfe_oauth_client_test.go`

```bash
❯ TESTARGS="-run TFEOAuthClient" TEST="./internal/provider" make testacc

TF_ACC=1 TF_LOG_SDK_PROTO=OFF go test ./internal/provider -v -run TFEOAuthClient -timeout 15m
=== RUN   TestAccTFEOAuthClientDataSource_basic
--- PASS: TestAccTFEOAuthClientDataSource_basic (4.98s)
=== RUN   TestAccTFEOAuthClientDataSource_findByID
--- PASS: TestAccTFEOAuthClientDataSource_findByID (3.28s)
=== RUN   TestAccTFEOAuthClientDataSource_findByName
--- PASS: TestAccTFEOAuthClientDataSource_findByName (3.34s)
=== RUN   TestAccTFEOAuthClientDataSource_findByServiceProvider
--- PASS: TestAccTFEOAuthClientDataSource_findByServiceProvider (3.10s)
=== RUN   TestAccTFEOAuthClientDataSource_missingParameters
--- PASS: TestAccTFEOAuthClientDataSource_missingParameters (0.13s)
=== RUN   TestAccTFEOAuthClientDataSource_missingOrgWithName
--- PASS: TestAccTFEOAuthClientDataSource_missingOrgWithName (0.13s)
=== RUN   TestAccTFEOAuthClientDataSource_missingOrgWithServiceProvider
--- PASS: TestAccTFEOAuthClientDataSource_missingOrgWithServiceProvider (0.12s)
=== RUN   TestAccTFEOAuthClientDataSource_sameName
--- PASS: TestAccTFEOAuthClientDataSource_sameName (2.34s)
=== RUN   TestAccTFEOAuthClientDataSource_noName
--- PASS: TestAccTFEOAuthClientDataSource_noName (1.94s)
=== RUN   TestAccTFEOAuthClientDataSource_sameServiceProvider
--- PASS: TestAccTFEOAuthClientDataSource_sameServiceProvider (2.37s)
=== RUN   TestAccTFEOAuthClient_basic
--- PASS: TestAccTFEOAuthClient_basic (2.75s)
=== RUN   TestAccTFEOAuthClientWithOrganizationScoped_basic
--- PASS: TestAccTFEOAuthClientWithOrganizationScoped_basic (2.52s)
=== RUN   TestAccTFEOAuthClient_rsaKeys
--- PASS: TestAccTFEOAuthClient_rsaKeys (2.71s)
PASS
```

_Please run applicable acceptance tests locally and include the output here. See [testing.md](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/testing.md) to learn how to run acceptance tests._

_If you are an external contributor, your contribution(s) will first be reviewed before running them against the project's CI pipeline._

```
$ TESTARGS="-run TestAccTFEWorkspace" make testacc

...
```

### CODE REVIEW

File: internal/provider/resource_tfe_oauth_client.go [hunk 1]
```diff
@@ -157,7 +158,7 @@ func resourceTFEOAuthClientCreate(d *schema.ResourceData, meta interface{}) erro
 	if serviceProvider == tfe.ServiceProviderAzureDevOpsServer {
 		options.PrivateKey = tfe.String(privateKey)
 	}
-	if serviceProvider == tfe.ServiceProviderBitbucketServer {
+	if serviceProvider == tfe.ServiceProviderBitbucketServer || serviceProvider == tfe.ServiceProviderBitbucketDataCenter {
```
Comment (laurenolivia) [inline internal/provider/resource_tfe_oauth_client.go:161]: It looks like you cannot call this endpoint to create the service you're adding in this PR. See [docs](https://developer.hashicorp.com/terraform/cloud-docs/api-docs/oauth-clients#create-an-oauth-client)

Comment (zainq11) [reply internal/provider/resource_tfe_oauth_client.go:161]: I will double check which create endpoint is being called @laurenolivia . I was able to test the provider locally and create an `oauth_client` for a `bitbucket_data_center` service provider.
```terraform
resource "tfe_oauth_client" "test" {
  name = "bbdc-cli-tester"
  # organization     = "zain-tfe"
  api_url          = "https://bitbucket.dev.tf-vcs.hashicorp.services"
  http_url         = "https://bitbucket.dev.tf-vcs.hashicorp.services"
  key              = "my_key"
  secret           = "-----BEGIN RSA PRIVATE KEY-----XXXXX-----END RSA PRIVATE KEY-----\n"
  rsa_public_key   = "-----BEGIN PUBLIC KEY-----XXXXX-----END PUBLIC KEY-----\n"
  service_provider = "bitbucket_data_center"
}
```

Comment (zainq11) [reply internal/provider/resource_tfe_oauth_client.go:161]: PR for terraform cloud api docs: https://github.com/hashicorp/terraform-docs-common/pull/624

Comment (zainq11) [reply internal/provider/resource_tfe_oauth_client.go:161]: Cloud API docs approved. Thanks Lauren! 😄

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@
 ENHANCEMENTS:
 * Rebrand Terraform Cloud to HCP Terraform by @sebasslash [#1328](https://github.com/hashicorp/terraform-provider-tfe/pull/1328)
 * Adds `post_apply` to list of possible `stages` for Run Tasks by @carolinaborim [#1307](https://github.com/hashicorp/terraform-provider-tfe/pull/1307)
+* `r/tfe_oauth_client`: Add Bitbucket Data Center support with the `bitbucket_data_center` option for `service_provider` by @zainq11 [#1303](https://github.com/hashicorp/terraform-provider-tfe/pull/1304)
```
Comment (brandonc) [inline CHANGELOG.md:8]: Before merging, move this entry up to the UNRELEASED section

**Comment by laurenolivia** [comment]
@zainq11 ping me if you need another review

**Comment by zainq11** [comment]
> @zainq11 ping me if you need another review

@laurenolivia Could you please review when you have time. I have made changes to the `CHANGELOG.md` and as discussed in slack, I have also run tests in `resource_tfe_oauth_client_test.go` which touch the updated resource. Please let me know if there is anything else I need to test as well. Happy to do that!

```bash
❯ TESTARGS="-run TFEOAuthClient" TEST="./internal/provider" make testacc

TF_ACC=1 TF_LOG_SDK_PROTO=OFF go test ./internal/provider -v -run TFEOAuthClient -timeout 15m
=== RUN   TestAccTFEOAuthClientDataSource_basic
--- PASS: TestAccTFEOAuthClientDataSource_basic (4.98s)
=== RUN   TestAccTFEOAuthClientDataSource_findByID
--- PASS: TestAccTFEOAuthClientDataSource_findByID (3.28s)
=== RUN   TestAccTFEOAuthClientDataSource_findByName
--- PASS: TestAccTFEOAuthClientDataSource_findByName (3.34s)
=== RUN   TestAccTFEOAuthClientDataSource_findByServiceProvider
--- PASS: TestAccTFEOAuthClientDataSource_findByServiceProvider (3.10s)
=== RUN   TestAccTFEOAuthClientDataSource_missingParameters
--- PASS: TestAccTFEOAuthClientDataSource_missingParameters (0.13s)
=== RUN   TestAccTFEOAuthClientDataSource_missingOrgWithName
--- PASS: TestAccTFEOAuthClientDataSource_missingOrgWithName (0.13s)
=== RUN   TestAccTFEOAuthClientDataSource_missingOrgWithServiceProvider
--- PASS: TestAccTFEOAuthClientDataSource_missingOrgWithServiceProvider (0.12s)
=== RUN   TestAccTFEOAuthClientDataSource_sameName
--- PASS: TestAccTFEOAuthClientDataSource_sameName (2.34s)
=== RUN   TestAccTFEOAuthClientDataSource_noName
--- PASS: TestAccTFEOAuthClientDataSource_noName (1.94s)
=== RUN   TestAccTFEOAuthClientDataSource_sameServiceProvider
--- PASS: TestAccTFEOAuthClientDataSource_sameServiceProvider (2.37s)
=== RUN   TestAccTFEOAuthClient_basic
--- PASS: TestAccTFEOAuthClient_basic (2.75s)
=== RUN   TestAccTFEOAuthClientWithOrganizationScoped_basic
--- PASS: TestAccTFEOAuthClientWithOrganizationScoped_basic (2.52s)
=== RUN   TestAccTFEOAuthClient_rsaKeys
--- PASS: TestAccTFEOAuthClient_rsaKeys (2.71s)
PASS
```

**Pull request merged**